frontend/src/metabase/collections/components/CollectionContent/CollectionItemsTable.tsx

```diff
@@ -0,0 +1,209 @@
+import type { Dispatch } from "metabase-types/store";
+
+import {
+  CollectionEmptyContent,
+  CollectionTable,
+} from "./CollectionContent.styled";
+
+const PAGE_SIZE = 25;
+const ALL_MODELS: CollectionItemModel[] = [
+  "dashboard",
+  "dataset",
+  "card",
+  "metric",
+  "snippet",
+  "collection",
+];
+
+const wrapCollectionItemList = (
+  itemList: CollectionItem[],
+  dispatch: Dispatch,
+) => {
+  return (itemList ?? []).map(object => {
+    const entity = entityForObject(object);
+    if (entity) {
+      return entity.wrapEntity(object, dispatch);
+    } else {
+      console.warn("Couldn't find entity for object", object);
+      return object;
+    }
+  });
+};
+
+export type CollectionItemsTableProps = {
+  collectionId: CollectionId;
+} & Partial<{
+  bookmarks: Bookmark[];
+  clear: () => void;
+  collection: Collection;
+  createBookmark: CreateBookmark;
+  databases: Database[];
+  deleteBookmark: DeleteBookmark;
+  getIsSelected: (item: CollectionItem) => boolean;
+  handleCopy: (items: CollectionItem[]) => void;
+  handleMove: (items: CollectionItem[]) => void;
+  hasPinnedItems: boolean;
+  loadingPinnedItems: boolean;
+  models: CollectionItemModel[];
+  pageSize: number;
+  selectOnlyTheseItems: (items: CollectionItem[]) => void;
+  selected: CollectionItem[];
+  toggleItem: (item: CollectionItem) => void;
+  onClick: (item: CollectionItem, event: any) => void;
+}>;
+
+export const CollectionItemsTable = ({
+  collectionId,
+  collection,
+  getIsSelected,
+  selectOnlyTheseItems,
+  databases,
+  bookmarks,
+  createBookmark,
+  deleteBookmark,
+  loadingPinnedItems,
+  hasPinnedItems,
+  selected,
+  toggleItem,
+  clear,
+  handleMove,
+  handleCopy,
+  pageSize = PAGE_SIZE,
+  models = ALL_MODELS,
+  onClick,
+}: CollectionItemsTableProps) => {
+  const [unpinnedItemsSorting, setUnpinnedItemsSorting] =
+    useState<SortingOptions>({
+      sort_column: "name",
+      sort_direction: SortDirection.Asc,
+    });
+
+  const { handleNextPage, handlePreviousPage, setPage, page, resetPage } =
+    usePagination();
+
+  useEffect(() => {
+    if (collectionId) {
+      resetPage();
+    }
+  }, [collectionId, resetPage]);
+
+  const handleUnpinnedItemsSortingChange = useCallback(
+    (sortingOpts: SortingOptions) => {
+      setUnpinnedItemsSorting(sortingOpts);
+      setPage(0);
+    },
+    [setPage],
+  );
+
+  const unpinnedQuery = {
+    collection: collectionId,
+    models,
+    limit: pageSize,
+    offset: pageSize * page,
+    ...(isRootTrashCollection(collection)
+      ? {}
+      : { pinned_state: "is_not_pinned" }),
+    ...unpinnedItemsSorting,
+  };
+
+  const {
+    data = [],
+    isLoading: loadingUnpinnedItems,
+    metadata,
+  } = useSearchListQuery({
+    // @ts-expect-error SearchRequest[models] is SearchModel[] but we use CollectionItem[] for this query
+    query: unpinnedQuery,
+  });
+
+  const dispatch = useDispatch();
+
+  const collectionItemList = wrapCollectionItemList(data, dispatch);
```

Comment:
I'll move these into a `useMemo` and do a few optimizations while this is being reviewed. Also I'm going to try using [Million Link](https://million.dev/) locally to see if I can speed things up.